sanic/server/websockets/impl.py

```diff
@@ -0,0 +1,776 @@
+    pings: Dict[bytes, asyncio.Future]  # Dict[bytes, asyncio.Future[None]]
+    pings: Dict[bytes, asyncio.Future]  # Dict[bytes, asyncio.Future[None]]
+    conn_mutex: asyncio.Lock
+    recv_lock: asyncio.Lock
+    process_event_mutex: asyncio.Lock
+    can_pause: bool
+    data_finished_fut: Optional[
+        asyncio.Future
+    ]  # Optional[asyncio.Future[None]]
+    pause_frame_fut: Optional[asyncio.Future]  # Optional[asyncio.Future[None]]
+    connection_lost_waiter: Optional[
+        asyncio.Future
+    ]  # Optional[asyncio.Future[None]]
+    keepalive_ping_task: Optional[asyncio.Task]
+    auto_closer_task: Optional[asyncio.Task]
+
+    def __init__(
+        self,
+        connection,
+        max_queue=None,
+        ping_interval: Optional[float] = 20,
+        ping_timeout: Optional[float] = 20,
+        close_timeout: Optional[float] = 10,
+        loop=None,
+    ):
+        self.connection = connection
+        self.io_proto = None
+        self.loop = None
+        self.max_queue = max_queue
+        self.close_timeout = close_timeout
+        self.ping_interval = ping_interval
+        self.ping_timeout = ping_timeout
+        self.assembler = WebsocketFrameAssembler(self)
+        self.pings = {}
+        self.conn_mutex = asyncio.Lock()
+        self.recv_lock = asyncio.Lock()
+        self.process_event_mutex = asyncio.Lock()
+        self.data_finished_fut = None
+        self.can_pause = True
+        self.pause_frame_fut = None
+        self.keepalive_ping_task = None
+        self.auto_closer_task = None
+        self.connection_lost_waiter = None
+
+    @property
+    def subprotocol(self):
+        return self.connection.subprotocol
+
+    def pause_frames(self):
+        if not self.can_pause:
+            return False
+        if self.pause_frame_fut is not None:
+            return False
+        if self.loop is None or self.io_proto is None:
+            return False
+        if self.io_proto.transport is not None:
+            self.io_proto.transport.pause_reading()
+        self.pause_frame_fut = self.loop.create_future()
+        return True
+
+    def resume_frames(self):
+        if self.pause_frame_fut is None:
+            return False
+        if self.loop is None or self.io_proto is None:
+            error_logger.debug(
+                "Websocket attempting to resume reading frames, but connection is gone."
+            )
+            return False
+        if self.io_proto.transport is not None:
+            self.io_proto.transport.resume_reading()
+        self.pause_frame_fut.set_result(None)
+        self.pause_frame_fut = None
+        return True
+
+    async def connection_made(self, io_proto: SanicProtocol, loop=None):
+        if loop is None:
+            try:
+                loop = getattr(io_proto, "loop")
+            except AttributeError:
+                loop = asyncio.get_event_loop()
+        self.loop = loop
+        self.io_proto = io_proto  # this will be a WebSocketProtocol
```

Comment:
This comment is redundant

Reply:
Yep. That was a reminder for me during dev. It can be removed.